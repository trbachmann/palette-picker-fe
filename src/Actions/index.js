export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const addProjectPalettes = (projectPalettes, id) => ({
  type: 'ADD_PROJECT_PALETTES',
  id,
  projectPalettes
})

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  isLoading: bool
});

export const setPaletteColors = (colors) => ({
  type: 'SET_PALETTE_COLORS',
  palette: colors
});

export const setCurrentExpandedProject = (id) => ({
  type: 'SET_CURRENT_EXPANDED_PROJECT',
  id
})
export const addProject = (name, id) => ({
  type: 'ADD_PROJECT',
  name,
  id
});

export const addPaletteToProject = (projectId, palette) => ({
  type: 'ADD_PALETTE',
  projectId,
  palette
});

export const deletePalette = (paletteId, projectId) => ({
  type: 'DELETE_PALETTE',
  paletteId,
  projectId
});

export const deleteProject = (projectId) => ({
  type: 'DELETE_PROJECT',
  projectId
});

export const expandMiniPalette = (colors) => ({
  type: 'EXPAND_MINI_PALETTE',
  colors
});

export const replacePalette = (palette) => ({
  type: 'REPLACE_PALETTE',
  palette
});

export const updateProjectName = (name) => ({
  type: 'UPDATE_PROJECT_NAME',
  name
});