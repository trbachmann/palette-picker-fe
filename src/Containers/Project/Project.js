import React, { Component } from 'react';
import MiniPaletteArea from '../MiniPaletteArea/MiniPaletteArea';

export class Project extends Component{
  constructor() {
    super();
    this.state = {
      expanded: false
    }
  }

  render() {
    return(
      <div>
        <div>
          <p>{this.props.name}</p>
          <i class="fas fa-pencil-alt"></i>
          <i class="far fa-trash-alt"></i>
        </div>
        {
          this.state.expanded &&
          <MiniPaletteArea/>
        }
      </div>
    )
  }
}

export default Project;