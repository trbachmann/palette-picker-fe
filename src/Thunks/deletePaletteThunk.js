import { toggleLoading, setError, deletePalette } from '../Actions';
import { fetchData, createOptions } from '../Utils/fetch';

export const deletePaletteThunk = (paletteId) => {
  return async (dispatch) => {
    const url = `/palettes/${paletteId}`;
    const options = createOptions('DELETE');
    try {
      dispatch(toggleLoading(true));
      await fetchData(url, options);
      dispatch(toggleLoading(false));
      dispatch(deletePalette(paletteId));
    } catch(error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message));
    }
  }
}