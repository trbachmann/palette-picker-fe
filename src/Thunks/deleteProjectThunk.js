import { toggleLoading, setError, deleteProject } from '../Actions';
import { fetchData, createOptions } from '../Utils/fetch';

export const deleteProjectThunk = (id) => {
  return async (dispatch) => {
    const url = `/projects/${id}`;
    const options = createOptions('DELETE');
    try {
      dispatch(toggleLoading(true));
      await fetchData(url, options);
      dispatch(toggleLoading(false));
      dispatch(deleteProject(id));
    } catch(error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message));
    }
  }
}