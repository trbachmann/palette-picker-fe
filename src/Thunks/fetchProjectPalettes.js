import { addProjectPalettes, setError, toggleLoading } from '../Actions';
import { fetchData } from '../Utils/fetch';

export const fetchProjectPalettes = (id) => {
  return async (dispatch) => {
    try{
      dispatch(toggleLoading(true));
      const projectPalettes = await fetchData(`/projects/${id}/palettes`);
      dispatch(toggleLoading(false));
      dispatch(addProjectPalettes(projectPalettes, id))
    } catch (error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message));
    }
  }
};