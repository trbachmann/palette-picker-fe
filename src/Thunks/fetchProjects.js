import { toggleLoading, setError, setProjects } from '../Actions';
import { fetchData } from '../Utils/fetch';

export const fetchProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(toggleLoading(true));
      const projects = await fetchData('/projects');
      dispatch(toggleLoading(false));
      dispatch(setProjects(projects));
    } catch (error) {
      dispatch(setError(error));
    }
  }
};