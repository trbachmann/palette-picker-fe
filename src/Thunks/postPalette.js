import { toggleLoading, setError, addPaletteToProject } from '../Actions';
import { fetchData, createOptions } from '../Utils/fetch';

export const postPalette = (projectId, palette) => {
  return async (dispatch) => {
    const url = '/palettes';
    const options = createOptions('POST', { ...palette, project_id: projectId });
    try {
      dispatch(toggleLoading(true));
      const paletteId = await fetchData(url, options);
      dispatch(toggleLoading(false));
      dispatch(addPaletteToProject(projectId, {...palette, id: paletteId.id }));
    } catch (error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message))
    }
  }
} 