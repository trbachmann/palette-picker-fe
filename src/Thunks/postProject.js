import { toggleLoading, setError, addProject } from '../Actions';
import { fetchData, createOptions } from '../Utils/fetch';

export const postProject = (projectName) => {
  return async (dispatch) => {
    const url = '/projects';
    const options = createOptions('POST', { name: projectName });
    try {
      dispatch(toggleLoading(true));
      const projectId = await fetchData(url, options);
      dispatch(toggleLoading(false));
      dispatch(addProject(projectName, projectId.id));
    } catch(error) {
      dispatch(toggleLoading(false));
      dispatch(setError(error.message))
    }
  }
}