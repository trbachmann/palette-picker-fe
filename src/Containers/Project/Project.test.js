import React from 'react';
import { shallow } from 'enzyme';
import { Project } from './Project';
import { fetchProjectPalettes } from '../../Thunks/fetchProjectPalettes';
import { deleteProjectThunk } from '../../Thunks/deleteProjectThunk';
import { setCurrentExpandedProject } from '../../Actions';
import { mapDispatchToProps } from './Project';
import { mapStateToProps } from './Project';

jest.mock('../../Thunks/fetchProjectPalettes');
jest.mock('../../Thunks/deleteProjectThunk');

describe('Project', () => {
  describe('Project component', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<Project/>);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot when expanded is true', () => {
      wrapper.setState({ expanded: true });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with the currentExpandedProject', () => {
      const mockState = {
        isLoading: false,
        error: '', 
        currentExpandedProject: 1
      };
      const expected = {
        currentExpandedProject: 1
      };
      const result = mapStateToProps(mockState);
      expect(result).toEqual(expected);
    })
  })

  describe('mapDispatchToProps', () => {
    let mockDispatch;

    beforeEach(() => {
      mockDispatch = jest.fn();
    });

    it('should call dispatch with fetchProjectPalettes', () => {
      const thunkToDispatch = fetchProjectPalettes(mockDispatch);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchProjectPalettes();
      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);
    })

    it('should call dispatch with setCurrentExpandedProject', () => {
      const actionToDispatch = setCurrentExpandedProject(1);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentExpandedProject(1);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with deleteProjectThunk', () => {
      const thunkToDispatch = deleteProjectThunk(mockDispatch);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.deleteProjectThunk();
      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);
    });
  });
});