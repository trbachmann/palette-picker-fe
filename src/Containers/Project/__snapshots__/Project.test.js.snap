// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Project component should match the snapshot 1`] = `
<div>
  <div
    className="Project--div--name"
  >
    <h3
      onClick={[Function]}
    />
    <i
      className="far fa-trash-alt"
      onClick={[Function]}
    />
  </div>
  <MiniPaletteArea />
</div>
`;

exports[`Project Project component should match the snapshot when expanded is true 1`] = `
<div>
  <div
    className="Project--div--name"
  >
    <h3
      onClick={[Function]}
    />
    <i
      className="far fa-trash-alt"
      onClick={[Function]}
    />
  </div>
  <MiniPaletteArea />
</div>
`;
